### PR TITLE
Login Endpoint

### LINKED ISSUE #27: POST /api/login
Allow a user to login with a username an password. Use bcrypt to check against the hashed password on the DB

### PR DESCRIPTION
**Pull request by dyazdani**
## Description
Login endpoint that makes a POST request to /api/account/login.

## Related Issue

Closes #27 

## Checklist
- [x] Created tests which fail without the change (if possible)
- [x] Confirmed all tests are passing

## How Has This Been Tested?

Tested with Postman and Jest. Tested that a 200 response was made and that string was returned in the token property of the response body.
Screen Shot 2023-10-09 at 10 39 45
Screen Shot 2023-10-09 at 11 49 34
Screen Shot 2023-10-09 at 11 48 30

### CODE REVIEW

File: __tests__/api/account.test.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        await pool.end()
+    })
+
+    describe('POST api/account/login', () => {
+
+        beforeEach( async () => {
+            await createTables(pool)
+        })
+
+        it('should return a JSON Web Token with valid id and password', async () => {
```
Comment (paloobi) [inline __tests__/api/account.test.js:30]: Recommend: add a test for the negative - that it does not send back a token for invalid id & password

File: __tests__/api/account.test.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+            const bilbo = getBilbo();            
+            const customer = await request.post('/api/account/register').send(bilbo);
```
Comment (paloobi) [inline __tests__/api/account.test.js:32]: Do not use the API to register, that makes your login test dependent on your register logic. Use the DB adapter instead to create a customer object

File: __tests__/api/account.test.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            const id = customer.body.customer.id;
+            const response = await request.post('/api/account/login').send({id: id, password: bilbo.password});
+            expect(response.status).toBe(200);
+            expect(response.body.customer.full_name).toBe(bilbo.fullName);
+            expect(response.body).toHaveProperty('token');
+            expect(typeof response.body.token === 'string').toBe(true);
```
Comment (paloobi) [inline __tests__/api/account.test.js:38]: Remember to check it's not an empty string:
```suggestion
            expect(typeof response.body.token).toBe('string');
            expect(response.body.token).toBeTruthy()
```

File: server/api/account.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
+import express from 'express';
+const bcrypt = require('bcrypt');
+const jwt = require('jsonwebtoken');
+require('dotenv').config()
```
Comment (paloobi) [inline server/api/account.ts:4]: This needs to be moved much higher in your app otherwise dotenv will only be configured in this file? Move to your express app index file

Comment (paloobi) [reply server/api/account.ts:4]: Other things commonly stored in .env include DB url, JSON secrets (which you will need anywhere that you need to verify them), and hostname / port

File: server/api/account.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+const { getPool, createCustomer, getCustomerById} = require('../db');
+const pool = getPool();
+
+const saltRounds = 10;
```
Comment (paloobi) [inline server/api/account.ts:9]: This is a constant, make it screaming snake case
```suggestion
const SALT_ROUNDS = 10;
```

File: server/api/account.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+const accessTokenSecret = process.env.ACCESS_TOKEN_SECRET;
```
Comment (paloobi) [inline server/api/account.ts:11]: Same - const string should be all caps snake
```suggestion
const { ACCESS_TOKEN_SECRET }= process.env.ACCESS_TOKEN_SECRET;
```

File: server/api/account.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+        next(e);
+    }
+})
+
+// POST /api/account/login
+
+accountRouter.post('/login', async (req, res, next): Promise<void> => {
+    try {
+        const customer = await getCustomerById(pool, req.body.id);
+        const isValid = await bcrypt.compare(req.body.password, customer.password);
```
Comment (paloobi) [inline server/api/account.ts:35]: Remember to check if the user doesn't exist!
```suggestion
        const customer = await getCustomerById(pool, req.body.id);
        if (!customer) {
            res.status(404).send(...)
        }
        const isValid = await bcrypt.compare(req.body.password, customer.password);
```

File: server/api/account.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+        if (isValid) {
+            console.log('Password is valid')
+
+            // JSON Web Token returned to client
+            const token = jwt.sign({
+                username: customer.fullName,
+                id: customer.id,
+            }, accessTokenSecret);
```
Comment (paloobi) [inline server/api/account.ts:44]: Why are you generating a username on this object? These users don't have a username. Email feels like the closest choice for a unique username that won't change)
```suggestion
            const token = jwt.sign({
                email: customer.email,
                id: customer.id
            }, accessTokenSecret);
```

File: server/api/account.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+            // res.json({token});
+            delete customer.password;
+            res.send({customer, token}) // TODO: is this the proper way to return the token with customer details?
```
Comment (paloobi) [inline server/api/account.ts:47]: It's up to you and what you want to have on your front-end. I think it's reasonable to include both the customer and the token in the response, though some APIs will only return the token. This way you wouldn't need a second round-trip to get the customer data (eg. for the account page, or to display the logged in user's name in the top nav), so I think this is nice

**Review by paloobi — APPROVED** [review]

File: server/api/account.ts (hunk 6 above)
Comment (dyazdani) [reply server/api/account.ts:11]: How is destructuring working here? Isn't `ACCESS_TOKEN_SECRET` a string?

File: server/api/account.ts (hunk 8 above)
Comment (dyazdani) [reply server/api/account.ts:44]: username was in your instructions on the GitHub issue, but email would work fine.

File: server/api/account.ts (hunk 6 above)
Comment (paloobi) [reply server/api/account.ts:11]: Oh sorry I meant to delete it from the part after the assignment.
```
const { ACCESS_TOKEN_SECRET }= process.env;
```
